src/btree/bt_walk.c

```diff
@@ -424,10 +425,14 @@ __tree_walk_internal(WT_SESSION_IMPL *session, WT_REF **refp, uint64_t *walkcntp
                 if (LF_ISSET(WT_READ_NO_WAIT) && current_state != WT_REF_MEM)
                     break;
             } else if (current_state == WT_REF_DELETED) {
+                /* Ensure we have a snapshot to check visibility or we only check global visibility.
+                 */
+                WT_ASSERT(session,
+                  LF_ISSET(WT_READ_VISIBLE_ALL) || F_ISSET(session->txn, WT_TXN_HAS_SNAPSHOT));
```

Comment:
Let's assert this at the top of the function so we don't do it on every deleted ref?

Reply:
I tried that and it hits too many errors. Let me try again.